### PR TITLE
[improvement](nereids) support extract from disjunction in join on condition

### PR DESCRIPTION
**Pull request by feiniaofeiafei**
```sql
select * from test_like1 t1 inner join test_like2 t2 on t1.a=t2.a and (t2.a=10 && t1.a=8 || t1.a=9);
```
support extract t1.a in (8,9) from   t1.a=t2.a and (t2.a=10 && t1.a=8 || t1.a=9), and push to scan.

### CODE REVIEW

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run external

**Comment by feiniaofeiafei** [comment]
run p0

**Comment by feiniaofeiafei** [comment]
run p0

**Review by 924060929 — DISMISSED** [review]

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ExtractSingleTableExpressionFromDisjunction.java [hunk 1]
```diff
@@ -50,24 +52,44 @@
  * 3. In old optimizer, there is `InferFilterRule` generates redundancy expressions. Its Nereid counterpart also need
  * `RemoveRedundantExpression`.
  * <p>
- * TODO: This rule just match filter, but it could be applied to inner/cross join condition.
  */
-public class ExtractSingleTableExpressionFromDisjunction extends OneRewriteRuleFactory {
+public class ExtractSingleTableExpressionFromDisjunction implements RewriteRuleFactory {
+    private static final ImmutableSet<JoinType> ALLOW_JOIN_TYPE = ImmutableSet.of(JoinType.INNER_JOIN,
+            JoinType.LEFT_OUTER_JOIN, JoinType.RIGHT_OUTER_JOIN, JoinType.LEFT_SEMI_JOIN, JoinType.RIGHT_SEMI_JOIN,
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ExtractSingleTableExpressionFromDisjunction.java:58]: why not support full outer join?

Comment (feiniaofeiafei) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ExtractSingleTableExpressionFromDisjunction.java:58]: full outer join on condition predicate cannot be pushdown

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run cloud_p0

**Review by morrySnow — APPROVED** [review]

**Review by xzj7019 — APPROVED** [review]

**Pull request merged**